### PR TITLE
Open log by db_number

### PR DESCRIPTION
**Pull request by Kaszub09**
Added method to open specific log instance using db_number.

`open_log` method opens only the last log found, and `desc` doesn't have to be unique - so if there are more than two logs with same `object`/`subobject`/`desc`, you can't specify which one to open. I think `db_number` is better for this purpose, since it uniquely identifies log instance.
E.g. for every run of a background job you could store just `db_number` and be able to get log of specific run. Alternative would be to try and make `desc` unique across runs, which might not always make sense.

### CODE REVIEW

File: src/zcl_logger_factory.clas.abap [hunk 1]
```diff
@@ -48,6 +48,16 @@ CLASS zcl_logger_factory DEFINITION
...
         VALUE(r_display_profile) TYPE REF TO zif_logger_display_profile.
 
+
+    "! Reopens specific log instance.
+    CLASS-METHODS open_log_by_db_number
+      IMPORTING
+        db_number        TYPE balognr
+        settings         TYPE REF TO zif_logger_settings OPTIONAL
+      RETURNING
+        VALUE(logger_if) TYPE REF TO zif_logger.
```
Comment (AlexandreHT) [inline src/zcl_logger_factory.clas.abap:58]: Could you use r_log in order to be consistent with create_log and open_log?

**Comment by mbtools** [comment]
Why do we need a new method? 

How about passing `lognumber` as optional parameter to `open_log`. Then pass `lognumber` to `BAL_FILTER_CREATE` in `i_t_lognumber`. 

Also `object` and `subobjects` optional so you can pass just the log number.

File: src/zcl_logger_factory.clas.abap (hunk 1 above)
Comment (Kaszub09) [reply src/zcl_logger_factory.clas.abap]: Sure

**Comment by Kaszub09** [comment]
@mbtools 
It's possible, but I don't like that solution, because then you end up with another problem, common in SAP, in that you have multiple optional parameters and have to guess, often by trial and error, which one you need pass, what combination is valid, and how they interact with each other. 

For instance, what should happen if i pass both object/subobject and db_number? What if db_number have different object than the one passed? And now it's possible to pass only object or even no parameters at all, which doesn't make sense. And so on. You could try to describe it all in documentation, but you may miss something or forgot to update it. At least now it's clear what exactly is required.

Maybe you could abstract some common code to another private method, but that's about it.

That's how I view it, but if others also prefer having one method, then sure. :)

**Comment by AlexandreHT** [comment]
@Kaszub09: @mbtools and I exchanged our views and we agree that a new methods is the way to go, but you need to factor the common code of ```open_log``` and this new method in order to avoid code duplication.
Thanks!

**Comment by Kaszub09** [comment]
At first I wanted to just call `open_log_by_db_number` from `open_log`, but turns out calling `BAL_DB_LOAD` twice with same number doesn't fill headers table second time (just why, SAP?).

So to avoid getting handle directly from database I've extracted to another methods both filtering, which returns handle even if called twice, and log creation by handle returned from filter.

I've also restored version because I've realized it probably should be changed in pull request, since there can be several at time and they can be merged in any order.

**Comment by AlexandreHT** [comment]
Code looks good to me, thanks for your work. I would still have :
- 1 request: could you add your comment about ```BAL_DB_LOAD```'s behaviour into the code. It'll be helpful someday
- 1 suggestion: considering ```get_log_headers``` returns either 0, 1 or n entries, would you consider renaming the method into ```find_log_headers``` or ```search_log_headers```? It's common (though not that much in ABAP) that ```get``` is reserved to methods that return either one entry or an exception.

**Review by AlexandreHT — APPROVED** [review]

**Pull request merged**